cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -915,6 +915,68 @@ public long getProgramRunCount(ProgramId programId) throws NotFoundException {
     });
   }
 
+  /**
+   * Method to add old run count info
+   *
+   * @param maxRows max batch size to fetch
+   */
+  public void upgrade(int maxRows) {
+    // If upgrade is already complete, then simply return.
+    if (isUpgradeComplete()) {
+      LOG.debug("Run count is already upgraded.");
+      return;
+    }
+
+    boolean activeUpgradeComplete = false;
+    boolean computeComplete = false;
+    while (!isUpgradeComplete()) {
```

Comment:
this while loop contains too many things and should be broken up. 

Writing the start time does not need to be in a loop and marking upgrade as complete should not be done in a loop.

The other parts should be in their own loops instead of using booleans to control which ones run.